nab/detectors/htmcore/pandaComm/pandaServer.py

```diff
@@ -0,0 +1,303 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Created on Mon Aug 26 09:22:19 2019
+
+@author: osboxes
+"""
+
+import socket, pickle, struct
+from socket import error as SocketError
+import threading
+from pandaComm.dataExchange import ServerData, CLIENT_CMD, SERVER_CMD
+import numpy
+
+verbosityLow = 0
+verbosityMedium = 1
+verbosityHigh = 2
+FILE_VERBOSITY = (
+    verbosityMedium
+)  # change this to change printing verbosity of this file
+
+
+def printLog(txt, verbosity=verbosityLow):
+    if FILE_VERBOSITY >= verbosity:
+        print(txt)
+
+
+def PackData(cmd, data):
+    d = [cmd, data]
+    # Pickle the object and send it to the server
+    # protocol must be specified to be able to work with py2 on server side
+    rawData = pickle.dumps(d, protocol=2)
+
+    if len(rawData) % 4096 == 0:  # if length of data is multiple of chunck size
+        # increase the data by 1 byte to prevent it - it causes problems in recv function
+        # on the client side - because client doesn't know how long data to expect
+        if isinstance(data, ServerData):
+            data.compensateSize.append(1)  # increase size by some dummy bytes
+            d = [cmd, data]
+            rawData = pickle.dumps(d)  # , protocol=2)
+        else:
+            printLog("Packed data is multiple of chunck size, but not known instance")
+
+    return rawData
+
+
+def send_one_message(sock, data):
+    length = len(data)
+    sock.sendall(struct.pack("!I", length))
+    sock.sendall(data)
+
+
+def recv_one_message(sock):
+    lengthbuf = sock.recv(4)
+    length, = struct.unpack("!I", lengthbuf)
+    return sock.recv(length)
+
+
+class PandaServer:
```

Comment:
no problem to include at all, but we can install from a git repo with pip, so `HTMpandaVis` could be added directly to NAB requirements. Would that help? (or is this file customized for the usecase?)

Reply:
This file is not customized, it is general. It would be nice to setup this :thumbsup:

I am not sure how to do it tho.. should i make special repo that will contain just pandaComm python package ?

Reply:
> This file is not customized, it is general. It would be nice to setup this

great! Let's see how problematic that would be. 

> .. should i make special repo that will contain just pandaComm python package ?

I would rather avoid further fragmentation. Already this problem spans htmcore, your repo, and NAB. 

Looking at HTMpandaVis install: 

> Install prerequisities & clone pandaVis
sudo apt-get install python3-tk
git clone https://github.com/htm-community/HTMpandaVis.git
python3 -m pip install -r requirements.txt

Could you add `setup.py` to the root of your repo, which installs the core/server (=general) components? That way users who use it with their own system (like here, NAB) can just: 
- modify their specific code (that will always be needed). 
  - Another BTW, if the code seems stable, we can setup an API that will be used by HTMcore, and ideally by the community who want to visualize the networks. (Is there some API like that in general? for neural net visualizations?) 
- install the server components from `pip` easily

Reply:
Ok so i added setup.py into HTMpanda repo, that installs the PandaComm package that the server (user custom scripts) will use. The readme in panda repo is also updated. I think this is the best solution for now. No need to use pip (test.pypi) since the client app is run directly as python code and this i want to keep for now. Becouse of that, user need to git clone repo anyways, so running setup.py for installing pandaComm package is not additional big effort.